Project Principles, Readme & Contributing

liamdennehy opened the pull request:
In order to help smooth out discussions I thought it useful to capture some [guiding principles](https://github.com/liamdennehy/http-signatures/blob/contributing/CONTRIBUTING.md#principles) for this specification. Hopefully these can be agreed on and used as a reference point for future and existing issues. This also captures the "tribal knowledge" formally, which I think is useful. Right now these are buried deep in some lengthy issue discussions.

To illustrate:

The issues on permitting or denying multiple "Signature" headers cannot be resolved until we first agree multiple header instances are ok in any case, and how to sign multiple instance headers depends on answering the same question. Since the HTTP RFCs are not particularly clear, we need to either be strict or tolerant in our interpretation, but it should at least be consistent. I have formulated the first bullet under the principle "[Simple and Compatible](https://github.com/liamdennehy/http-signatures/blob/contributing/CONTRIBUTING.md#simple-and-compatible)" with this in mind, hopefully in a way that does not cause controversy. 

I suggest opening new issues with the "project" tag for specific issues rather than going down the comments road here, and tagging this PR in those issues.

liamdennehy committed:
CONTRIBUTING: First Draft

liamdennehy committed:
README: First Draft

liamdennehy committed:
README: Fixups

liamdennehy committed:
CONTRIBUTING: fixups

liamdennehy committed:
CONTRIBUTING: typo

liamdennehy committed:
CONTRIBUTING: Simplicity and accuracy

liamdennehy committed:
CONTRIBUTING: Clarify out-of-band preferences

liamdennehy committed:
CONTRIBUTING: What You See is What You Sign

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+## Platform, Language & Transport neutrality
+
+This protocol is meant to provide a technical mechanism to produce and
+verify digital signatures of HTTP Messages in a platform-agnostic way
```
iiuc this protocol wants to sign `HTTP header fields`. This is different from the HTTP Message which is subject to  Transfer-Coding  and mangling of intermediaries.

liamdennehy replied:
See 7594164. Keen to not get overly specific in the guidelines, as this is covered in the spec.

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+  - If HTTP does or permits something, this protocol should allow it to be
+  signed.
+
+  - This means we also support things that present for legacy reasons or
```
Not sure. I think we should plan the protocol to evolve and deprecate feature. 
See https://tools.ietf.org/id/draft-thomson-postel-was-wrong-03.html

liamdennehy replied:
I put this in specifically to address e.g. multiple header instances. I'm keen to make sure we simply add security, and don't force implementations to change how they use HTTP in order to get there.

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+  be able to ignore any signature header safely.
+
+- The specification should avoid wherever possible any aspects requiring
+  negotiation or agreement between parties. Out-of-band specification of
```
Can you make an example please?

liamdennehy replied:
https://github.com/w3c-dvcg/http-signatures/issues/25, https://github.com/w3c-dvcg/http-signatures/issues/6, https://github.com/w3c-dvcg/http-signatures/issues/4, https://github.com/w3c-dvcg/http-signatures/issues/3

All involve extra round-trips and possible negotiation, while this principle expects these to be agreed beforehand, and out-of-band. Any time security parameters are negotiated we need to add defensive measures, which don't further the central purpose - add security.

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+# What You See is What You Sign
+
+- Signatures serve an immediate purpose to authenticate and verify a
+  HTTP message at the point it is received and processed, so the protocol
```
`the point it is received` is the User Agent? Every Intermediary receives and processes messages...
Do you mean the intended recipient of the signature?

liamdennehy replied:
All systems in the path, and eventually the target.
3231d95?

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+  should provide a mechanism to determine the timeliness of a signature
+
+- Signatures may serve an after-the-fact purpose to be able to verify that a
+  request or response was properly verified at some point in the past
```
Do you mean we need a proof of the *validation process*?

liamdennehy replied:
No, just that real-time is not the only use-case. e.g. a message may be in a non-repudiation store and require review months later. Can you suggest wording if this is unclear? Next paragraph has the real meat, this is just context.

ioggstream commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,71 @@
+
+- To serve both the above purposes, signatures should be verifiable given
+  the information available at that time. Signatures should not rely on
+  information not present in a given message except for the secrets used.
```
This means that the trust model should be set out-of-band. Is there something else we are saying there?

liamdennehy replied:
e.g. ``(request-target)`` is not transmitted in a response, so a response' signature that includes it would have to have that available too. Inside a single transaction it is available from the context (I as client just sent it, so I know the target, so I can use it to verify), but if the response alone is stored and requires verification later, the target info is not included so the signature cannot be reconstructed.

ioggstream commented on README.md:
```diff
@@ -0,0 +1,80 @@
+sgnatures in multipart messages using e.g. GPG.
+
+However there is no accepted mechanism to provide protection to the HTTP
+message as a whole, and especially within a single HTTP request/response
```
imho we are protecting part of the http message. 

we are not protecting the http message as a whole. If once we send a message, an intermediary
decides to apply transfer-coding and deliver that via multiple messages, he can do that.

Moreover we don't have a way to express that some headers MUST NOT be present, so the message is changed.

liamdennehy replied:
That's why I say "mechanism". This protocol cannot protect an entire message, but in conjunction with payload digests it can, which we should illustrate but not consider core.

ioggstream replied:
I'll try to PR next week: hope to clarify.

ioggstream reviewed (changes requested):
Stubbed some considerations.

liamdennehy committed:
CONTRIBUTING: Clarify header & payload protection approaches

liamdennehy committed:
CONTRIBUTING: Timeliness during processing, forwarding and receipt

Remove bias on server/client and acknowledge intermediate systems
(proxies etc)

liamdennehy commented:
I don't think this has value in being merged given the age and that this has now turned into a published IETF RFC. Happy to close without merge.
